service/qos/service_level_controller.cc

```diff
@@ -221,6 +221,20 @@ future<> service_level_controller::update_service_levels_from_distributed_data()
                 do_add_service_level(sl.first, sl.second).get();
             }
         });
+    }).handle_exception([this] (std::exception_ptr eptr) {
+        try {
+            std::rethrow_exception(eptr);
+        } catch (const broken_semaphore& ex) {
+            // Catch broken semaphore exception.
+            // If it was caused by early abort subscription, ignore the exception.
+            // Otherwise, rethrow it.
+
+            if (!*_early_abort_subscription) {
+                // Early abort subscription was fired
+                return make_ready_future();
+            }
+            std::rethrow_exception(std::move(eptr));
+        }
```

Comment:
I'm not a fan of this solution. It relies on non-local reasoning (it assumes that if there is a `broken_semaphore` exception and `!*_early_abort_subscription`, then it must be correlated).

It looks like we are using the `notifications_serializer` semaphore for two reasons:

1. To serialize operations on service levels,
2. To prevent more operations from being done on service levels when we are shutting down.

It currently fits purpose (1) well and we do want to serialize the current operations as we currently do (i.e. reloading SL state, adding/removing SLs when an incoming SL is not known). However, I think there is a problem with purpose (2). I think it would be much cleaner to add a `gate` and make it responsible for (2). You could remove the `.broken()` call in `do_abort()`, and only close the gate in `stop()`.

Reply:
A solution I like (don't know it it applies well here) is to use an update loop that waits on a condition variable and an abort_source. When the condition variable is signaled, it reads and applies the changes. The service waits to the update loop in its stop() method.

Reply:
This has the nice property that if the update loop is slow, then multiple updates are coalesced.

Reply:
The current update loop is being deprecated so we are not going to change the algorithm. Instead, the SL state update will be triggered by group0. I agree that it's a nice way to write an update loop, though.

The problem with the code in the PR right now is that the semaphore is broken when the global abort_source is being triggered. This semaphore is taken in the function that reloads the SL state and which is invoked by group0 code, so the exception from the broken semaphore breaks the group0 state machine.

I suggested splitting the semaphore into a semaphore + gate, but perhaps we could just "break" the semaphore in service_level_controller::stop()...? (AFAIK it is called after group0 is stopped, so it should work). I'm not familiar with using broken semaphores to control access to a component, though, so I suggested using a gate.

Reply:
I wasn't referring to the current update loop. Just a nice pattern to serialize and batch events.

Perhaps we can capture it in some template so it can be easily reused.